src/arkode/arkode_io.c

```diff
@@ -34,6 +35,347 @@
+
+  int num_bool_keys = 4;
+  const char* bool_keys[] = {"arkode.nonlinear",
+                             "arkode.clear_stop_time",
+                             "arkode.no_inactive_root_warn",
+                             "arkode.reset_accumulated_error"};
+  arkBoolSetFn bool_set[] = {ARKodeSetNonlinear,
+                             ARKodeClearStopTime,
+                             ARKodeSetNoInactiveRootWarn,
+                             ARKodeResetAccumulatedError};
+
+  int i, j, retval;
+  for (i = 1; i < argc; i++)
+  {
+    sunbooleantype arg_used = SUNFALSE;
+
+    /* skip command-line arguments that do not begin with "arkode." */
+    if (strncmp(argv[i], "arkode.", 7) != 0) { continue; }
+
+    /* check all "int" command-line options */
+    for (j = 0; j < num_int_keys; j++)
+    {
+      if (CheckAndSetIntArg(ark_mem, &i, argv, int_keys[j],
+                            int_set[j], &arg_used) != ARK_SUCCESS)
+      {
+        return ARK_ILL_INPUT;
+      }
+      if (arg_used) break;
+    }
+    if (arg_used) continue;
+
+    /* check all long int command-line options */
+    for (j = 0; j < num_long_keys; j++)
+    {
+      if (CheckAndSetLongArg(ark_mem, &i, argv, long_keys[j],
+                             long_set[j], &arg_used) != ARK_SUCCESS)
+      {
+        return ARK_ILL_INPUT;
+      }
+      if (arg_used) break;
+    }
+    if (arg_used) continue;
+
+    /* check all real command-line options */
+    for (j = 0; j < num_real_keys; j++)
+    {
+      if (CheckAndSetRealArg(ark_mem, &i, argv, real_keys[j],
+                             real_set[j], &arg_used) != ARK_SUCCESS)
+      {
+        return ARK_ILL_INPUT;
+      }
+      if (arg_used) break;
+    }
+    if (arg_used) continue;
+
+    /* check all bool command-line options */
+    for (j = 0; j < num_bool_keys; j++)
+    {
+      if (CheckAndSetBoolArg(ark_mem, &i, argv, bool_keys[j],
+                             bool_set[j], &arg_used) != ARK_SUCCESS)
+      {
+        return ARK_ILL_INPUT;
+      }
+      if (arg_used) break;
+    }
+    if (arg_used) continue;
+
+    /*** handle all remaining command-line options ***/
+
+    if (strcmp(argv[i], "arkode.interpolant_type") == 0)
+    {
+      i++;
+      retval = 1;
+      if (strcmp(argv[i], "ARK_INTERP_HERMITE") == 0)
+      {
+        retval = ARKodeSetInterpolantType(arkode_mem, ARK_INTERP_HERMITE);
+      }
+      else if (strcmp(argv[i], "ARK_INTERP_LAGRANGE") == 0)
+      {
+        retval = ARKodeSetInterpolantType(arkode_mem, ARK_INTERP_LAGRANGE);
+      }
+      else if (strcmp(argv[i], "ARK_INTERP_NONE") == 0)
+      {
+        retval = ARKodeSetInterpolantType(arkode_mem, ARK_INTERP_NONE);
+      }
+      if (retval != ARK_SUCCESS)
+      {
+        arkProcessError(ark_mem, ARK_ILL_INPUT, __LINE__, __func__, __FILE__,
+                        "error setting command-line argument: %s %s",
+                        argv[i-1], argv[i]);
+        return ARK_ILL_INPUT;
```

Comment:
It seems this function always returns `ARK_ILL_INPUT` on a failure. If the input was ok, but a function like `ARKodeSetInterpolantType` failed for a different reason, e.g., some internal interpolation memory was corrupted, it would be more informative to pass along that error.